src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,98 @@ def self.compile(compile_options)
       end
 
       def self.compile_ast(code, compile_options)
-        Puppet[:node_name_value] = compile_options['certname']
-
-        # Use the existing environment with the requested name
-        Puppet::Pal.in_environment(compile_options['environment'],
-                                   envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+        # If the request requires that bolt be loaded we assume we are in a properly 
+        # configured PE environment
+        if compile_options.dig('options', 'bolt')
+          require 'bolt'
+          require 'bolt/target'
+          require 'bolt/apply_inventory'
+          require 'bolt/apply_target'
+          require 'bolt/apply_result'
+          require 'addressable'
```

Comment:
Just to make sure I understand. We need 'bolt/apply_inventory', 'bolt/apply_target', and 'bolt/apply_result' to be loaded so that PCore types that may be in apply block ASTs are defined. But bolt/apply_target requires bolt/target to be loaded. And all of them require 'bolt' to have been loaded previously, but bolt doesn't require addressable? Is that correct?

Reply:
This seems largely correct. I did not attempt to *not* whole sale require bolt. I feel like bolt/apply_target should be requiring addressable (but currently does not https://github.com/puppetlabs/bolt/pull/1770/files#diff-452723efd86a7b4a811e7856a524a1d7R3). I think i could spend some more time investigating this.